## Recall o Sensibilidad 

El Recall, también conocido como Sensibilidad o Tasa Positiva Verdadera (TPR por sus siglas en inglés, True Positive Rate), es una métrica de desempeño utilizada en problemas de clasificación. Se enfoca en evaluar la capacidad de un modelo para identificar correctamente todos los casos positivos en el conjunto de datos.

Supongamos que tenemos un modelo de clasificación binaria que predice si un correo electrónico es spam o no spam. Tenemos un conjunto de 1000 correos electrónicos para evaluar el desempeño de nuestro modelo, y hemos obtenido las siguientes predicciones y etiquetas reales:

- Verdaderos Positivos (TP): 150 correos electrónicos fueron correctamente clasificados como spam.
- Verdaderos Negativos (TN): 700 correos electrónicos fueron correctamente clasificados como no spam (ham).
- Falsos Positivos (FP): 50 correos electrónicos fueron incorrectamente clasificados como spam cuando eran en realidad no spam (falsas alarmas).
- Falsos Negativos (FN): 100 correos electrónicos fueron incorrectamente clasificados como no spam cuando eran en realidad spam (correos importantes que se ignoraron).

In [1]:
import pandas as pd
from sklearn.metrics import recall_score

data = {
    'Prediccion': [1] * 150 + [0] * 700 + [1] * 50 + [0] * 100,
    'Real': [1] * 150 + [0] * 700 + [0] * 50 + [1] * 100
}
df = pd.DataFrame(data)
df

Prediccion  Real
0             1     1
1             1     1
2             1     1
3             1     1
4             1     1
..          ...   ...
995           0     1
996           0     1
997           0     1
998           0     1
999           0     1

[1000 rows x 2 columns]

In [2]:
recall = recall_score(df.Real, df.Prediccion)
print('Recall score:', recall)

Recall score: 0.6



### Recall = $\frac{{Verdaderos Positivos (TP)}}{{Verdaderos Positivos (TP) + Falsos Negativos (FN)}}$


### Recall = $\frac{150}{150 + 100} = \frac{150}{250} = 0.6$

Esto significa que el modelo tiene un Recall de 0.6, lo que indica que es capaz de identificar correctamente el 60% de todos los correos electrónicos que eran realmente spam.

![TPR](TPR.jpeg)

## Especificidad o TNR (Tasa Negativa Verdadera)

La especificidad (tasa de verdaderos negativos) mide la proporción de negativos que se identifican correctamente (es decir, la proporción de aquellos que no tienen la afección (no afectados) que se identifican correctamente como personas que no padecen la afección).

Usando el ejemplo anterior tenemos que:

In [6]:
TP = 150
TN = 700
FP = 50

# Calcular el TNR (Especificidad)
tnr = TN / (TN + FP)

print("TNR (Especificidad):", tnr)

TNR (Especificidad): 0.9333333333333333


### TNR = $\frac{{Verdadero Negativo (TN)}}{{Verdadero Negativo (TN) + Falsos Positivos (FP)}}$
### TNR = $\frac{{700}}{{700 + 50}} = \frac{700}{750} = 0.933333$

Esto significa que el modelo clasificó correctamente aproximadamente el 93.33% de los correos electrónicos no spam como no spam y cometió errores de falsa alarma en aproximadamente el 6.67% de los casos.

![TNR](TNR.png)